google-authenticator.php

```diff
@@ -161,12 +180,345 @@ function add_qrcode_script() {
+
+	// Site admin screen
+	add_submenu_page( 'options-general.php', esc_html__( 'Google Authenticator', 'google-authenticator' ), esc_html__( 'Google Authenticator', 'google-authenticator' ), 'manage_options', 'google_authenticator', array( $this, 'admin_setup_page' ) );
+
+	// Network admin screen
+	add_submenu_page( 'settings.php', esc_html__( 'Google Authenticator', 'google-authenticator' ), esc_html__( 'Google Authenticator', 'google-authenticator' ), 'manage_network_options', 'google_authenticator', array( $this, 'network_admin_setup_page' ) );
+}
+
+/**
+ * Determine if a user needs to setup authy 2fa
+ * @return bool
+ */
+function user_needs_to_setup_google_authenticator() {
+	$user = wp_get_current_user();
+	$enabled = trim(get_user_option( 'googleauthenticator_enabled', $user->ID ) ) === 'enabled';
+	if ( $enabled ) {
+		return false;
+	}
+
+	$must_signup = false;
+	$user_role = $user->roles[0];
+	$check_single_site_admin_options = true;
+
+	if ( is_multisite() ) {
+		$roles = get_site_option( 'googleauthenticator_mandatory_mfa_roles', array() );
+		if ( in_array( $user_role, $roles ) ) {
+			$must_signup = true;
+		}
+		$check_single_site_admin_options = '1' !== get_site_option( 'googleauthenticator_network_only' ) ;
+	}
+
+	if ( ! $must_signup && $check_single_site_admin_options ) {
+		$roles = get_option( 'googleauthenticator_mandatory_mfa_roles', array() );
+		if ( in_array( $user_role, $roles ) ) {
+			$must_signup = true;
+		}
+
+	}
+
+	return apply_filters( 'google_authenticator_needs_setup', $must_signup, $user );
+}
+
+/**
+ * Send users to the signup page if they must signup.
+ */
+function redirect_if_setup_required() {
+	if ( $this->user_needs_to_setup_google_authenticator() ) {
+		$screen = get_current_screen();
+		$pagename = 'admin_page_' . self::SETUP_PAGE;
+		if ( is_a( $screen, 'WP_Screen') && in_array( $screen->id, array( $pagename, 'profile' ) ) ) {
+			return;
+		}
+
+		// Some check against super admin so they can enable/disable the plugin.
+		$location = admin_url( 'admin.php?page=' . self::SETUP_PAGE );
+		wp_redirect( $location);
+		exit;
+	}
+}
+
+/**
+ * Save the GA secret if valid totp is provided
+ * @return void|WP_Error
+ */
+function save_submitted_setup_page() {
+	$secret = empty( $_POST['GA_secret'] ) ? false : sanitize_text_field( $_POST['GA_secret']);
+	$otp = empty( $_POST['GA_otp_code'] ) ? false : sanitize_text_field( $_POST['GA_otp_code']);
+	if ( ! strlen( $secret ) || ! strlen( $otp ) ) {
+		return;
+	}
+	if ( $timeslot = $this->verify( $secret, $otp ) ) {
+		global $user;
+		update_user_option( $user->ID, 'googleauthenticator_lasttimeslot', $timeslot, true );
+		update_user_option( $user->ID, 'googleauthenticator_secret', $secret, true );
+		update_user_option( $user->ID, 'googleauthenticator_enabled', 'enabled', true );
+		$location = admin_url( 'index.php?googleauthenticator=enabled' );
+		wp_redirect( $location );
```

Comment:
@ivankruchkoff This doesn't seem to be working correctly as before request enters here other part of WordPress already outputs HTML code

wp-admin/admin.php:212

```
                do_action( "load-{$page_hook}" );
		if (! isset($_GET['noheader']))
			require_once(ABSPATH . 'wp-admin/admin-header.php');

		/**
		 * Used to call the registered callback for a plugin screen.
		 *
		 * @ignore
		 * @since 1.5.0
		 */
		do_action( $page_hook );
```

HTML output is generated by

```
if (! isset($_GET['noheader']))
			require_once(ABSPATH . 'wp-admin/admin-header.php');
```

and out `user_setup_page -> $this->save_submitted_setup_page()` is being invoked by `do_action( $page_hook );`

I'd presonally move`$error = $this->save_submitted_setup_page();` to `load-admin_page_google_authenticator_user_page` action and I'd assign `new WP_Error( 'invalid-otp', esc_html__( "OTP code doesn't match supplied secret, please check you've configured Authenticator correctly.") );` to class property that would be checked in `user_setup_page` instead of `$error = $this->save_submitted_setup_page();`